src/gd_avif.c

```diff
@@ -0,0 +1,667 @@
+    if (rgb.pixels) {
+      avifRGBImageFreePixels(&rgb);
+    }
+
+    if (encoder) {
+      avifEncoderDestroy(encoder);
+    }
+
+    if (avifOutput.data) {
+      avifRWDataFree(&avifOutput);
+    }
+
+    return failed;
+}
+
+/*****************************************************************************
+ * 
+ *                            HELPER FUNCTIONS
+ * 
+ *****************************************************************************/
+
+/* Convert the quality param we expose to the quantity params used by libavif.
+   The *Quantizer* params values can range from 0 to 63, with 0 = highest quality and 63 = worst.
+   We make the scale 0-100, and we reverse this, so that 0 = worst quality and 100 = highest.
+
+   Values below 0 are set to 0, and values below MAX_QUALITY are set to MAX_QUALITY.
+*/
+
+static int quality2Quantizer(int quality) {
+  int clampedQuality = AVIF_CLAMP(quality, 0, MAX_QUALITY);
+
+  float scaleFactor = (float) AVIF_QUANTIZER_WORST_QUALITY / (float) MAX_QUALITY;
+
+  return round(scaleFactor * (MAX_QUALITY - clampedQuality));
+}
+
+/*
+  From gd_png.c:
+    convert the 7-bit alpha channel to an 8-bit alpha channel.
+    We do a little bit-flipping magic, repeating the MSB
+    as the LSB, to ensure that 0 maps to 0 and
+    127 maps to 255. We also have to invert to match
+    PNG's convention in which 255 is opaque. 
+*/
+
+static uint8_t convertTo8BitAlpha(uint8_t originalAlpha) {
+  return originalAlpha == 127 ? 
+        0 : 
+        255 - ((originalAlpha << 1) + (originalAlpha >> 6));
+}
+
+/*
+   As of February 2021, this algorithm reflects the latest research on how many tiles
+   and threads to include for a given image size.
+   This is subject to change as research continues.
+ */
+
+static void setEncoderTilesAndThreads(avifEncoder * encoder, avifRGBImage * rgb) {
+  int imageArea, tiles, tilesLog2;
+  int tileRowsLog2, tileColumnsLog2, maxThreads;
+
+  imageArea = rgb->width * rgb->height;
+
+  tiles = ceil(imageArea / MIN_TILE_AREA);
+  tiles = AVIF_MIN(tiles, MAX_TILES);
+  tiles = AVIF_MIN(tiles, MAX_THREADS);
+
+  tilesLog2 = floor(log10(tiles) / log10(2));
+
+  // If the image's width is greater than the height, use more tile columns
+  // than tile rows to make the tile size close to a square
+    
+  if (rgb->width >= rgb->height) {
+    encoder->tileRowsLog2 = tilesLog2 / 2;
+    encoder->tileColsLog2 = tilesLog2 - encoder->tileRowsLog2;
+  } else {
+    encoder->tileColsLog2 = tilesLog2 / 2;
+    encoder->tileRowsLog2 = tilesLog2 - encoder->tileColsLog2;
+  }
+
+  // It's good to have one thread per tile.
+  encoder->maxThreads = tiles;
+}
+
+/*
+   Check the result from an Avif function to see if it's an error.
+   If so, decode the error and output it, and return true.
+   Otherwise, return false.
+*/
+static avifBool isAvifError(avifResult result, char * msg) {
+  if (result != AVIF_RESULT_OK) {
+    gd_error("avif error: %s: %s", msg, avifResultToString(result));
+    return GD_TRUE;
```

Comment:
the return is `avifBool`, but you're using gd's `GD_TRUE` ... should be consistent

Reply:
True. I ended up changing all my GD_(TRUE|FALSE) to AVIF_(TRUE|FALSE), since there's no GD "boolean" type, and since the other instances were placed into libavif structures.